crates/primitives/src/log.rs

```diff
@@ -116,21 +116,49 @@ impl Log {
     }
 }
 
+impl<T> Log<T>
+where
+    for<'a> &'a T: Into<LogData>,
+{
+    /// Creates a new log.
+    #[inline]
+    pub const fn new_from_event_unchecked(address: Address, data: T) -> Self {
+        Self { address, data }
+    }
+
+    /// Creates a new log from an deserialized event.
+    pub fn new_from_event(address: Address, data: T) -> Option<Self> {
+        let this = Self::new_from_event_unchecked(address, data);
+        (&this.data).into().is_valid().then_some(this)
+    }
+
+    /// Reserialize the data.
+    #[inline]
+    pub fn reserialze(&self) -> Log<LogData> {
+        Log { address: self.address, data: (&self.data).into() }
+    }
+}
+
 #[cfg(feature = "rlp")]
-impl alloy_rlp::Encodable for Log {
+impl<T> alloy_rlp::Encodable for Log<T>
+where
+    for<'a> &'a T: Into<LogData>,
+{
     fn encode(&self, out: &mut dyn alloy_rlp::BufMut) {
+        let this = self.reserialze();
         let payload_length =
-            self.address.length() + self.data.data.length() + self.data.topics.length();
+            this.address.length() + this.data.data.length() + this.data.topics.length();
 
         alloy_rlp::Header { list: true, payload_length }.encode(out);
-        self.address.encode(out);
-        self.data.topics.encode(out);
-        self.data.data.encode(out);
+        this.address.encode(out);
+        this.data.topics.encode(out);
+        this.data.data.encode(out);
     }
 
     fn length(&self) -> usize {
+        let this = self.reserialze();
         let payload_length =
-            self.address.length() + self.data.data.length() + self.data.topics.length();
+            this.address.length() + this.data.data.length() + this.data.topics.length();
         payload_length + alloy_rlp::length_of_length(payload_length)
```

Comment:
this might need to be `payload_length + alloy_rlp::length_of_length(payload_length) + 1`? or `Header { list: true, payload_length }.length() + payload_length`, would check length assert tho

Reply:
length assert in latest commit works fine

Reply:
ah, forgot that the method for header just inlined `length_of_length`